stdlib/builtins.pyi

```diff
@@ -1747,19 +1746,45 @@ if sys.version_info >= (3, 10):
     class EncodingWarning(Warning): ...
 
 if sys.version_info >= (3, 11):
-    _SplitCondition = Union[type[BaseException], tuple[type[BaseException], ...], Callable[[BaseException], bool]]
+    _BaseExceptionT_co = TypeVar("_BaseExceptionT_co", bound=BaseException, covariant=True)
+    _BaseExceptionT = TypeVar("_BaseExceptionT", bound=BaseException)
+    _ExceptionT_co = TypeVar("_ExceptionT_co", bound=Exception, covariant=True)
+    _ExceptionT = TypeVar("_ExceptionT", bound=Exception)
 
-    class BaseExceptionGroup(BaseException):
-        def __new__(cls: type[Self], __message: str, __exceptions: Sequence[BaseException]) -> Self: ...
+    class BaseExceptionGroup(BaseException, Generic[_BaseExceptionT_co]):
+        def __new__(cls: type[Self], __message: str, __exceptions: Sequence[_BaseExceptionT_co]) -> Self: ...
         @property
         def message(self) -> str: ...
         @property
-        def exceptions(self) -> tuple[BaseException, ...]: ...
-        def subgroup(self: Self, __condition: _SplitCondition) -> Self | None: ...
-        def split(self: Self, __condition: _SplitCondition) -> tuple[Self | None, Self | None]: ...
-        def derive(self: Self, __excs: Sequence[BaseException]) -> Self: ...
+        def exceptions(self) -> tuple[_BaseExceptionT_co | BaseExceptionGroup[_BaseExceptionT_co], ...]: ...
+        @overload
+        def subgroup(
+            self, __condition: type[_BaseExceptionT] | tuple[type[_BaseExceptionT], ...]
+        ) -> BaseExceptionGroup[_BaseExceptionT] | None: ...
+        @overload
+        def subgroup(self: Self, __condition: Callable[[_BaseExceptionT_co], bool]) -> Self | None: ...
+        @overload
+        def split(
+            self: Self, __condition: type[_BaseExceptionT] | tuple[type[_BaseExceptionT], ...]
+        ) -> tuple[BaseExceptionGroup[_BaseExceptionT] | None, Self | None]: ...
+        @overload
+        def split(self: Self, __condition: Callable[[_BaseExceptionT_co], bool]) -> tuple[Self | None, Self | None]: ...
+        def derive(self: Self, __excs: Sequence[_BaseExceptionT_co]) -> Self: ...
```

Comment:
This is a bit of a tricky one. If I read PEP 654 correctly, the `__excs` provided don't need to have anything in common with the exceptions of `self`. So a "correct" annotation would be:

```python
def derive(self: Self, __excs: Sequence[_BaseExceptionT]) -> Self[_BaseExceptionT]: ...
```

But I don't think that's possible. So I'm fine with this, until someone comes up with a better alternative.

Reply:
Yet another use case for https://github.com/python/typing/issues/548...